contracts/airdropsSMCampaign/Existing1155NftDropSMCampaign.sol

```diff
@@ -0,0 +1,180 @@
+contract Existing1155NftDropSMCampaign is AirdropInfoSMCampaign, AirdropMerkleProof, IERC1155Receiver, Ownable {
+    IERC1155 public immutable rewardedNft;
+    IERC1155 public immutable rewardToken;
+    uint256 public immutable rewardTokenId;
+    uint256 public immutable tokensPerClaim;
+    uint256 public immutable totalAirdropAmount;
+
+    bool public airdropFunded = false;
+    uint256 public airdropFundBlockTimestamp;
+    address internal airdropFundingHolder;
+    address public immutable airBroFactoryAddress;
+
+    event Claimed(address indexed claimer);
+    event AirdropFunded(address contractAddress);
+    event MerkleRootChanged(bytes32 merkleRoot);
+    // event isEligibleForReward(bool);
+
+    error NotOwner();
+    error AirdropStillInProgress();
+    error AlreadyRedeemed();
+    error AlreadyFunded();
+    error InsufficientAmount();
+    error InsufficientLiquidity();
+    error AirdropExpired();
+    error NotAdmin();
+    error NotEligible();
+
+
+    mapping(address => bool) public hasClaimed;
+
+    // The root hash of the Merle Tree we previously generated in our JavaScript code. Remember
+    // to provide this as a bytes32 type and not string. Ox should be prepended.
+    bytes32 public merkleRoot;
+
+    uint256 public immutable airdropDuration;
+    uint256 public immutable airdropStartTime;
+    uint256 public immutable airdropFinishTime;
+
+    modifier onlyAdmin(){
+        if(msg.sender != IAirBroFactory(airBroFactoryAddress).admin()) revert NotAdmin();
+        _;
+    }
+
+    constructor(
+        address _rewardedNft,
+        address _reward1155Nft,
+        uint256 _tokensPerClaim,
+        uint256 _tokenId,
+        uint256 _totalAirdropAmount,
+        uint256 _airdropDuration,
+        address _airBroFactoryAddress
+    ) {
+        rewardedNft = IERC1155(_rewardedNft);
+        tokensPerClaim = _tokensPerClaim;
+        totalAirdropAmount = _totalAirdropAmount;
+        rewardToken = IERC1155(_reward1155Nft);
+        rewardTokenId = _tokenId;
+        airdropDuration = _airdropDuration * 1 days;
+        airdropStartTime = block.timestamp;
+        airdropFinishTime = block.timestamp + airdropDuration;
+        airBroFactoryAddress = _airBroFactoryAddress;
+
+    }
+
+    /// @notice Sets the merkleRoot - can only be done if admin (different from the contract owner)
+    /// @param _merkleRoot - The root hash of the Merle Tree
+    function setMerkleRoot(bytes32 _merkleRoot) external onlyAdmin {
+        merkleRoot = _merkleRoot;
+        emit MerkleRootChanged(_merkleRoot);
+    }
+
+    /// @notice Allows the airdrop creator to provide funds for airdrop reward
+    function fundAirdrop() external {
+        if (rewardToken.balanceOf(msg.sender, rewardTokenId) < totalAirdropAmount) revert InsufficientAmount();
+        if (airdropFunded) revert AlreadyFunded();
+        rewardToken.safeTransferFrom(msg.sender, address(this), rewardTokenId, totalAirdropAmount, "");
```

Comment:
use Checks Effects Interaction Patern
https://medium.com/returnvalues/smart-contract-security-patterns-79e03b5a1659#:~:text=Checks%20Effects%20Interaction%20Pattern%20is,feature%20in%20a%20certain%20function.

Reply:
Yes, that's how I understood his fdbk as well